source/common/router/upstream_request.cc

```diff
@@ -794,17 +837,25 @@ void UpstreamRequest::DownstreamWatermarkManager::onAboveWriteBufferHighWatermar
   // The downstream connection is overrun. Pause reads from upstream.
   // If there are multiple calls to readDisable either the codec (H2) or the underlying
   // Network::Connection (H1) will handle reference counting.
-  parent_.parent_.cluster()->trafficStats()->upstream_flow_control_paused_reading_total_.inc();
-  parent_.upstream_->readDisable(true);
+  if (!parent_.upstream_wait_for_response_headers_before_disabling_read_) {
+    parent_.parent_.cluster()->trafficStats()->upstream_flow_control_paused_reading_total_.inc();
+    parent_.upstream_->readDisable(true);
+  } else {
+    parent_.readDisableOrDefer(true);
+  }
 }
 
 void UpstreamRequest::DownstreamWatermarkManager::onBelowWriteBufferLowWatermark() {
   ASSERT(parent_.upstream_);
 
   // One source of connection blockage has buffer available. Pass this on to the stream, which
   // will resume reads if this was the last remaining high watermark.
-  parent_.parent_.cluster()->trafficStats()->upstream_flow_control_resumed_reading_total_.inc();
-  parent_.upstream_->readDisable(false);
+  if (!parent_.upstream_wait_for_response_headers_before_disabling_read_) {
+    parent_.parent_.cluster()->trafficStats()->upstream_flow_control_resumed_reading_total_.inc();
+    parent_.upstream_->readDisable(false);
```

Comment:
WDYT of moving this in the readDisableOrDefer function to keep the complexity in one place?